### PR TITLE
Added documentation for the resurrected blockchain.address.* methods

### PR DESCRIPTION
**Pull request by cculianu**
Hi @dagurval, as per our discussions I am documenting the resurrection of the `blockchain.address.*` methods.

I have this already implemented in my `dev` branch of Fulcrum and have tested it extensively.  A new version of Fulcrum 1.0.5 will be released soon that offers these resurrected methods.

I updated `protocol-changes`, `protocol-methods`, and `protocol-removed` to
reflect their new (old) documentation and their status as revived in
version 1.4.3 of the protocol.

Note I lack the tools on my system (and was a bit lazy, tbh) to actually run `make` to build the documentation. I did, however, use an online `rst` editor: https://livesphinx.herokuapp.com/. This editor seems good so I am optimistically hoping there are no render errors in my changes.

Would you mind also building these docs just to check?  In particular I added a **Note** box to `blockchain.address.subscribe`  which I feel needs double-checking that it renders right in the final built docs.

All apologies for not having the tools installed to build these docs -- I will figure it out but I just wanted this PR submitted today to get it out.

### CODE REVIEW

File: protocol-methods.rst [hunk 1]
```diff
@@ -2,6 +2,149 @@
...
+Return the confirmed and unconfirmed balances of a Bitcoin Cash address.
+
+**Signature**
+
+  .. function:: blockchain.address.get_balance(address)
+  .. deprecated:: 1.2 removed in version 1.3, re-added in version 1.4.3
+
+  * *address*
+
+    The address as a Cash Address (with or without prefix) or a Base58 string.
```
Comment (dagurval) [inline protocol-methods.rst]: Can we please drop base58 support? I mean, you can support it in Fulcrum, but I don't think we should force deprecated stuff on everyone in the protocol.

File: protocol-methods.rst [hunk 2]
```diff
@@ -2,6 +2,149 @@
...
+  signature is
+
+  .. function:: blockchain.address.subscribe(address, status)
+
+  .. note:: The address returned back to the client when notifying of status
+            changes will be in the same encoding and syle as was provided when
+            subscribing. In effect, a whitespace-stripped version of the address
+            string that the client provided will be sent back to the client when
+            notifying, in order to make it easier for clients to track the
+            notification.
```
Comment (dagurval) [inline protocol-methods.rst:142]: This seems unecessarily hard to implement. Can't we just make this lower case representation of cashaddr with prefix? No ambiguity.

File: protocol-methods.rst [hunk 3]
```diff
@@ -1016,8 +1159,8 @@ Only the first :func:`server.version` message is accepted.
 
 **Example**::
 
-  server.version("Electrum 3.0.6", ["1.1", "1.2"])
+  server.version("Electron Cash 3.3.6", ["1.2", "1.4"])
```
Comment (dagurval) [inline protocol-methods.rst:1200]: Unrelated change

File: protocol-methods.rst (hunk 1 above)
Comment (cculianu) [reply protocol-methods.rst]: Ok. How shall we word this then? 


>The address as a Cash Address (with or without prefix).  Some server implementations MAY also support Legacy (base58) encoded addresses as well.

^ How does that sound?

File: protocol-methods.rst (hunk 2 above)
Comment (cculianu) [reply protocol-methods.rst:142]: Why be so restrictive though?  I see that as needlessly annoying... why?  The idea is to make the server as easy to use for JS kiddies... the more little nits it has the more annoying it is to use.

File: protocol-methods.rst (hunk 3 above)
Comment (cculianu) [reply protocol-methods.rst:1200]: So?  Let's just add it -- :) Can also add Electrs too here. :)

**Comment by cculianu** [comment]
Note I went ahead and added a new protocol method `blockchain.address.get_scripthash` so that some libs/authors can work with scripthashes anyway by translating whatever address they have to a scripthash first before issuing `blockchain.scripthash.*` methods.

It's documented and I pushed a commit ...

I also edited some more stuff in the docs which you may or may not find unrelated.

File: protocol-methods.rst (hunk 1 above)
Comment (cculianu) [reply protocol-methods.rst]: Ok, I just pushed another commit just now.  Note there are 2 new commits now for you to review.

**Comment by cculianu** [comment]
@dagurval Ping?  I need a place to point to this new API.  I may just put up a temp page with these built docs in the meantime...

File: protocol-methods.rst (hunk 3 above)
Comment (dagurval) [reply protocol-methods.rst:1200]: Looks like a "this is the reference implementation" change to me, bad vibes. We have enough of wannabe reference implementations. How about making these client strings fictional clients?

**Comment by dagurval** [comment]
Pong, I'm on it.

**Comment by dagurval** [comment]
What do you think of these changes? https://github.com/dagurval/electrum-cash-protocol/pull/13/commits/651bb4f0e016cdce524c6c745127aab358fa1347

* Adds `base58` to `servers.features`
* Has the Fulcrum section removed (sorry!) -- rationale: It was basically just saying "Fulctrum does this :point_up:"

File: protocol-methods.rst (hunk 2 above)
Comment (dagurval) [reply protocol-methods.rst:142]: OK

**Comment by dagurval** [comment]
I opened #13 to rebase for you in case you got annoyed that I pulled out the misc fixes and merged already

File: protocol-methods.rst (hunk 3 above)
Comment (cculianu) [reply protocol-methods.rst:1200]: We might as well advertise our software dude.  I can add ElectrsCash in there too.  Stop being so rigid.  Nobody besides Electron Cash is even using this protocol now and it only connects to 3 servers.  Jesus.

Anyway I pushed a commit to satisfy you.

**Comment by cculianu** [comment]
I am not a fan of 651bb4f0e016cdce524c6c745127aab358fa1347.  I don't want to add this as a feature flag.  I want to support legacy 100%.  It was a concession for me to agree to make the protocol be vague and make it optional.

You should add legacy support.  It makes things easier for potential future clients.  People still use legacy, like it or not.

There is no reason to be draconian with such a 2% coin.  We want to embrace users, devs, everybody.  Making it restrictive for no reason is not ideal.

**Comment by cculianu** [comment]
Let me say this: I already implemented all the methods described, including legacy support.  Now, you're asking me to go back and create a feature flag because you don't like legacy.  Perhaps the feature flag should be "nolegacy" for servers that don't support legacy. That would only be fair. The early bird gets the worm.

**Comment by cculianu** [comment]
Also the **entire philosophy** behind adding these methods is to make it easier for script kiddies to just get things done.  Being restrictive and not supporting legacy puts a burden on clients that want to offer legacy to their users to now actually use a bitcoin lib to translate, thereby basically detracting from **the whole reason** why we added these methods in the first place -- **to not require a bitcoin lib** to talk to the server!

*sigh*

**Comment by dagurval** [comment]
> Let me say this: I already implemented all the methods described, including legacy support. Now, you're asking me to go back and create a feature flag because you don't like legacy. Perhaps the feature flag should be "nolegacy" for servers that don't support legacy. That would only be fair. The early bird gets the worm.

I'm not asking you to do anything. Why are you even asking for feedback if you've already got the perfect implementation?

**Comment by dagurval** [comment]
Can we have a second opinion here @imaginaryusername @monsterbitar ?

**Comment by imaginaryusername** [comment]
It seems like the contention is just over whether the server advertises base58 support or not, which is a useful but very small detail - why don't we just say "the server _can_ advertise this support via a base58 flag"? That way people who don't want to advertise it carries on, but people who do want to advertise explicitly can still do so via a common path.

**Comment by cculianu** [comment]
@imaginaryusername The contention is over not supporting legacy.  The whole point of this feature is to make it easier to write a simple client. By deciding to not support legacy the usefulness of this feature is diminished.

**Comment by imaginaryusername** [comment]
@cculianu i mean the two versions both have legacy support as optional, they only differ in whether the support is advertised by server or not...

**Comment by imaginaryusername** [comment]
*In other words, if we just say "server can/may" advertise, then Fulcrum doesn't need to do anything to stay in spec, but other implementations who want to advertise their base58 support explicitly can still do so via a common path instead of inventing their own.

**Comment by cculianu** [comment]
So rather than take the 25 minutes it would take to implement legacy -- someone now has to go ahead and add a feature flag nobody on the planet cares about? 

It's almost as easy to just add legacy support.  Why all the unnecessary unnecessaryness?! Jesus man.

**Comment by cculianu** [comment]
Look -- the spec should require legacy.  Period.

**Comment by imaginaryusername** [comment]
Again, the actual difference is not whether someone implements legacy - your version doesn't strictly require legacy support too! If the point of contention is over requiring legacy support or not it'll be over another set of text.

**Comment by cculianu** [comment]
>  your version doesn't strictly require legacy support too! 

Incorrect. It did until @dagurval had me change it.  But I had time to think about it and I think it's a bad idea to not require it.

Again, the spec should require legacy. Period.

**Comment by imaginaryusername** [comment]
So the contention is over https://github.com/dagurval/electrum-cash-protocol/pull/11/commits/ea49396dd56c97398d28674907289ef34b089248 instead of https://github.com/dagurval/electrum-cash-protocol/commit/651bb4f0e016cdce524c6c745127aab358fa1347 ?

**Comment by cculianu** [comment]
The contention is that the spec should require legacy and not having it do so diminishes from the whole point of this feature.

**Comment by dagurval** [comment]
Any script kiddie can convert to cashaddr. Here's how to do it in javascript:

```
const bchaddr = require("bchaddrjs");
const cashaddr = bchaddr.toCashAddress('1B9UNtBfkkpgt8kVbwLN9ktE62QKnMbDzR');
```

The real value in `blockchain.address.*` is not to have to deal with scripthash.

**Comment by imaginaryusername** [comment]
Okay, that a discussion I thought closed with https://github.com/dagurval/electrum-cash-protocol/commit/ea49396dd56c97398d28674907289ef34b089248. 

I agree that requiring it is cleaner and people might have strong opinion about cleaniness. But then the client already needs to read `protocol_max` in order to know whether they can use `blockchain.address.*` at all, it does not seem like a huge jump to also find Fulcrum from `server_version` to identify base58 support...

**Comment by cculianu** [comment]
Explain why you don't want to support it @dagurval ?  Is it just laziness?

**Comment by imaginaryusername** [comment]
Let's postpone the fight for 1.4.4 and get the objectively useful `blockchain.address` in first, shall we? :)

**Comment by dagurval** [comment]
> Explain why you don't want to support it @dagurval ? Is it just laziness?

Once you make it mandatory, it's hard to get it out again. If it's optional, there is an opt-out. People are still loosing money to base58 addresses, I'd like to not contribute to them being around forever.

**Comment by cculianu** [comment]
Ok, so we just agree to disagree here. I think server APIs should be as useful as possible -- I think the whole point of these methods is to just allow 1-off queries from software with 0 libs.  You feel it's contributing to the entropy and chaos to implement it.  I can't force you to implement it. I think for now just leaving it as-is, optional, no feature flag necessarily is probably the least controversial thing then.

**Comment by monsterbitar** [comment]
> You should add legacy support. It makes things easier for potential future clients. People still use legacy, like it or not.

I can see that for those that do use it, removing it is problematic. However, for all those who don't use it, enforcing support for it increases barriers to entry and can result in loss of users just as well.

My preferred outcome would be for the specification to **recommend** supporting base58 as well (a recommendation that can be removed in the future), but mark it clear in the documentation that this feature cannot be relied upon going forward.

_That said, I have no intent to ever use base58 again unless required to by an external party._

I don't know if the feature flag provides much value in this context but it seems that the cost to add it  should be low. Anyone wanting to support base58 to begin with won't have an excessive amount of extra work to also add the feature flag.
